src/ripple/app/misc/SHAMapStore.h

```diff
@@ -66,6 +67,37 @@ class SHAMapStore
 
     /** Returns the number of file descriptors that are needed. */
     virtual int fdRequired() const = 0;
+
+    /** The minimum ledger to try and maintain in our database.
+
+        This defines the lower bound for attempting to acquire historical
+        ledgers over the peer to peer network.
+
+        With online_delete,this value is governed by the following
+        circumstances:
+
+        - With advisory_delete enabled and online_delete having been executed
+          previously, this value is set to one past the value that is allowed to
+          be deleted. In other words, anything greater than what can be deleted
+          should be acquired from the network and then retained. If nothing has
+          yet been deleted, then do not rely upon the value of what is allowed
+          to be deleted because it could cause fetching of more history than
+          configured by ledger_history.
```

Comment:
By default, can_delete is initially 0, so without this limit and with the new changes, any new system with advisory_delete and online_delete will backfill to full history.

For the user circumstance you describe, I think they could set ledger_history initially if they want a bunch of ledgers, then turn it down later and restart.

Also, I think that, in the scenario you describe, the previous behavior to this feature, the backfilling you describe only occurred after online_delete ran at least once, since there was no lastRotated value yet.

Anyway, it's bad to use the can_delete value because can_delete is often 0, which corresponds to full history / unbounded history retrieval.

Reply:
> By default, can_delete is initially 0, so without this limit and with the new changes, any new system with advisory_delete and online_delete will backfill to full history.

Good point. In digging through the code again, I realized that if the user sets `can_delete always`, on the next restart, we'll get an overflow initializing `minimumOnline_` to 0, which will then cause it to be initialized with `minSqlSeq()`.... So that's not so bad, but in the process of figuring that out, I changed an existing config to enable `advisory_delete`. It had rotated previously, so `minimumOnline_` got initialized to 1.

My point is that there are a lot of edge cases.

I have a slightly more modest suggestion than my earlier suggestion: https://github.com/ximinez/rippled/commit/803820a584ffd6716a24398dde45608c61ac0282
Basically instead of tracking the minimum to fetch, it keeps the last to not fetch, and uses '>' instead of '>=' when making the decision in `shouldAcquire`. I think this will help correct this handful of off-by-one errors, plus that potential overflow and edge case.

> Also, I think that, in the scenario you describe, the previous behavior to this feature, the backfilling you describe only occurred after online_delete ran at least once, since there was no lastRotated value yet.

I don't think that's true with the current release version. It's only restriction was that `can_delete > 0`.
* https://github.com/ripple/rippled/blob/master/src/ripple/app/ledger/impl/LedgerMaster.cpp#L1713-L1714
* https://github.com/ripple/rippled/blob/master/src/ripple/app/ledger/impl/LedgerMaster.cpp#L1877-L1878

I think it's good to use `can_delete` in most cases, because it gives the user more fine-grained control in most circumstances. I do agree that it shouldn't be used if a rotation has never been done, or in the case where the user has never set it - i.e. it's still 0.

Reply:
Good catch about the can_delete. I fixed that and the other possible problems arising from using can_delete as the floor by not using can_delete at all upon startup. Instead, it's always safe to set minimumOnline_ as the minimum persisted, regardless of online_delete, advisory_delete, etc. Once online_delete runs again, that value will be bumped up accordingly. This is a reduction in moving parts. Since can_delete is dynamically user configurable, it's open to any number of weird circumstances.

About the minimumOnline vs onlineCutoff, I'd prefer to keep things as they are. minimumOnline is more descriptive of what it represents, whereas onlineCutoff is one fewer than the minimum to keep online. Or it's the maximum to not necessarily keep online. I think the corner cases you've identified are fixed by removing can_delete from the picture.

Reply:
I just pushed updated comment describing this. It's a lot simpler.

Reply:
Simplified further. minimumOnline_ is only set/modified during run of online_delete.

Reply:
Here's a summary of the behavior now, and perhaps something that can be used to update the documentation (@mDuo13).

Historical ledger acquisition is based first on ledger_history. If a ledger is within configured ledger_history, then the server will attempt to retrieve it. If not, the ledger will be retrieved if it is >= the minimum ledger persisted in the SQL database, *or* if online_delete is currently running, then > the highest ledger that is in the process of being deleted. If no data has been persisted, then the additional check is not performed. In all cases, the server will attempt to retrieve ledgers within its ledger_history range.

I'm sure the wording can be improved, but that's the new behavior and I hope it's easier to reason about at this point.